datafusion/common/src/error.rs

```diff
@@ -131,6 +131,10 @@ pub enum DataFusionError {
     /// Errors from either mapping LogicalPlans to/from Substrait plans
     /// or serializing/deserializing protobytes to Substrait plans
     Substrait(String),
+
+    /// Errors for wrapping other errors.
```

Comment:
Why is this needed (compared to `Context` for example?) I don't understand why we need to keep the data wrappeed (why don't we just pass the underlying error back?)

Reply:
Because if we only pass the underlying error, we can't share the error (due to ownership constraints). For example, `RepartitionExec` needs to send errors to all output partitions. You can see this in the source code.
https://github.com/apache/datafusion/blob/8d006a287c653e8305827ad909c407baa7ed59fd/datafusion/physical-plan/src/repartition/mod.rs#L914-L921

So, how do we share the error?
1. `Implement Clone for DataFusionError`: This is not possible because nested Error types do not implement the`Clone` trait.
2. Pass `Arc<DatafusionError>` : I implemented it this way.

Reply:
If using `Context`, the underlying error is only passed as a single `Result` due to ownership constraints.

Reply:
I renamed it to `SharedError`

Reply:
Thank you @getChan --  As DataFusionError is used all over the place and is a very visible part of the API I was thinking it would be best if we can avoid adding too many variants. However, after playing with some other options (like deep cloning the error) I agree that this solution is the best.

to be consistent with the other variants which don't have a `Error` in their name, I renamed this to `DataFusionError::Shared`